Role based Access - New Teacher/Student Login & Reg - STILL needs fixing/testing

angadk4 opened the pull request:
Created a new system for RBAC. Essentially, individuals can now use a toggle at the top of the registration page to log in as either a teacher or student, which assigns the role of either "User" or "Teacher." Upon login, the system now uses an email query to verify whether the user is a teacher or student, and proceeds accordingly.

**Changes Made:**
- Added radio switch at the top of register.html page: can be on either the student or teacher position, but not both at once.
- Added relevant css styling for the role-switch in register.css.
- Updated relevant register.js code to add event listeners for the new role switch.
- Updated relevant register.js code to be able to retain position of switch upon failed attempt/redirect.
- Updated register.js code so that when in 'teacher' position, the grade field disappears and input for it is no longer needed for form submission.
- Updated register route in auth.py so the system now requests the role attribute, and based on "teacher" or "student" assigns a new User or Teacher model.
- Updated login route in auth.py so the system now uses the inputted email to determine whether the user is a teacher or student, and redirects accordingly.
- Added temporary teacher redirect page (we will decide what teachers should be able to do later).

**Notes/Issues:**
-  **IMPORTANT:** Under models.py, the Teacher class had yrs_specialization set to Nullable=False. This was causing issues as the registration for teachers does not ask for this field, so it should be nullable in reality. I am not sure of how to migrate the database, so I would appreciate if someone could migrate the database with my new changes or show me how.
- **IMPORTANT:** For some reason, line 68 in auth.py (_role = request.form.get("role"))_ is not functioning as intended. I am completely unsure why this issue exists, but it needs to be fixed for the RBAC system to work and be tested.

angadk4 committed:
Fixed questions.json error by correcting path in quizSubmit.py

angadk4 committed:
Added teacher/student switch to reg page

angadk4 committed:
Edited all relevant register.js functions to accomodate for new roleswitch

angadk4 committed:
Added all relevant backend logic for RBAC

angadk4 committed:
Resolved teacher submission issue on js side, POST issue still persists

angadk4 committed:
Other edits, teacher reg issue remains - needs database migration

angadk4 committed:
reverted role field to request after testing - irrelevant changes

angadk4 committed:
fixed issue preventing user signup - teacher issue still exists

alexchen2 commented on app/src/models.py:
```diff
@@ -117,7 +117,7 @@ class Teacher(UserMixin, db.Model):
     age = db.Column(db.Integer)
     registration_date = db.Column(db.DateTime, default=datetime.now(timezone(timedelta(hours=-5))))
-    yrs_experience = db.Column(db.Integer, nullable=False)
+    yrs_experience = db.Column(db.Integer, nullable=True)
```
Add this additional column in here for Teacher login: 
```python
failed_signin_attempts = db.Column(db.Integer, default=0)
```

alexchen2 replied:
Also, again: change every instance of `tid` or `'teacher.tid'` to `id` and `teacher.id` respectively (for Flask-Login to work).

alexchen2 commented on app/src/models.py:
```diff
@@ -117,7 +117,7 @@ class Teacher(UserMixin, db.Model):
+    yrs_experience = db.Column(db.Integer, nullable=True)
     specialization = db.Column(db.String(50), nullable=True)
     # students = db.relationship('User', secondary=teacher_student, backref='teachers')
     courses = db.relationship('Course', secondary=teacher_course, backref='teachers')
```
Overall comment on the idea of combining User and Teacher into one table `User` with specializations/sub-tables `Students` and `Teachers`; see further comments below for details.

alexchen2 commented on app/src/templates/register.html:
```diff
@@ -8,6 +8,13 @@
+        <label for="student">Student</label>
+        <input type="radio" id="teacher" name="role" value="teacher">
+        <label for="teacher">Teacher</label>
+    </div>    
```
Move all this to inside the form at line 26, after the `.error-msg` div; this will fix the issue regarding your `request.form.get("role")` line not working.

alexchen2 commented on app/src/static/css/register.css:
```diff
@@ -156,6 +156,36 @@ main a:hover::after {
+.role-switch {
+    display: flex;
+    justify-content: center;
+    margin: 20px 0;
```
Replace and insert the following to get a similar design to your original radio buttons:
- line 162 (in `.role-switch`, `margin: 20px 0;`): 
   replace with
   ```css
   left: 0;
   right: 0;
   top: -63px;  /* adjust this value as necessary for radio btn placement*/
   position: absolute;
   ```

- line 151 (in `#heading`, `margin-bottom: 20px`): 
   replace with `margin-bottom: 83px` (adjust as necessary for spacing around the radio btn)

- line 189 (in `#register-box`):
  add `position: relative` (defines the radio btn's absolute positioning in terms of the gray sign-up form box)

alexchen2 commented on app/src/static/js/register.js:
```diff
@@ -11,6 +11,7 @@ const dateToday = new Date();     // Latest date = today
 let yearEarliest = dateToday.getFullYear() - 101;
 let monthEarliest = dateToday.getMonth();
 let dayEarliest, dateEarliest;
+let selectedRole = document.querySelector('input[name="role"]:checked').value; // Stores position of role switch
```
_Note:_ All of the comments in the JS file are just small recommendations to shorten repetition in the code, but are not necessary at all.

With this addition of the radio buttons, `inputFields` now includes two extra NodeElements for the buttons. You can replace line 3 (const inputFields) for the following to separate the radio buttons from the rest of the input text/select fields:

```javascript
const inputFields = Array.from(document.querySelectorAll("form input, #form-grade"))
    .filter((inputField) => !(inputField.parentElement.classList.contains("role-switch")));

const roleSwitch = document.querySelectorAll("form .role-switch input");
```

alexchen2 commented on app/src/static/js/register.js:
```diff
@@ -91,6 +92,23 @@ function renderError() {
+
+    // Event listener for when the role is changed to 'student'
+    // This will re-evaluate the form submission criteria
+    document.getElementById("student").addEventListener("change", function() {
+        selectedRole = this.value;
+        toggleGradeField();
+        updateFieldState(); // Update the state of the form based on the new role
+        console.log(selectedRole)
+    });
+    // Event listener for when the role is changed to 'teacher'
+    document.getElementById("teacher").addEventListener("change", function() {
+        selectedRole = this.value;
+        toggleGradeField();
+        updateFieldState(); // Update the state of the form based on the new role
+        console.log(selectedRole)
+    });
```
With the edition of `roleSwitch` above, you can reduce this down to:

```js
roleSwitch.forEach((roleBtn) => {
    roleBtn.addEventListener("change", () => {
        selectedRole = this.value;
        toggleGradeField();
        updateFieldState(); // Update the state of the form based on the new role
        console.log(selectedRole)            
    })
})
```

alexchen2 commented on app/src/auth.py:
```diff
@@ -57,18 +65,23 @@ def register():
         age = calculate_age(date_of_birth)
-        user_email = User.query.filter_by(email=email).first()
-        user_name = User.query.filter_by(username=username).first()
+        if role == "teacher":
+            user_email = Teacher.query.filter_by(email=email).first()
+            user_name = Teacher.query.filter_by(username=username).first()
+        else:
+            user_email = User.query.filter_by(email=email).first()
+            user_name = User.query.filter_by(username=username).first()
```
There's currently an issue with registration between both roles, where a student and teacher are allowed to register with the same exact email and username, but only the teacher can log into their account—not the student. Check comments below for more details.

alexchen2 reviewed (changes requested):
Looked over your PR and the two issues you mentioned. For the problem with the database migration, I'm not too sure on how to do that myself—maybe @skydonline might be able to help you with that. Worst case scenario though, you can probably just delete the current instance/site.db file and generate a new one by running the server, since I think we currently don't have any important DB entries stored that we're using elsewhere. 

For the second issue with the `request.form.get("role")` line, this pops up since in register.html, the `.role-switch input` tags are not contained within the `.register-form` form element. Therefore, when the user presses the button attached to this form to submit the data to the backend, no data with the form name of "role" is sent, so the backend cannot find any "role" field in the accompanying FormData/dict object. I've made a couple of suggestions in the HTML and CSS files to maintain the page's previous appearance while moving the `.role-switch` div within the form, as well as a couple of small edits in related JS code for conciseness.

In terms of other points of note, please change every instance of "id" within models.py to simply `id`. Flask-Login requires the primary key of any entity being used in its user session to be labelled specifically as "id", and will crash the server otherwise.

alexchen2 commented:
Overall, there are two general things I wanted to ask you regarding the implementation of your RBAC model here: the mutual exclusivity of consumers being either Users or Teachers, and the design similarities of the Users & Teachers database tables. 

For the former, how are you setting up the distinction between consumers within these roles? Can a consumer be both a teacher and a student, or is it an "either-or", one-or-the-other situation? Looking at your `auth.py` login code at lines 26 to 31, it appears as if you're going for the latter option. You attempt to query one entry in both User and Teacher with the given login email and expect one of them to be `None`:
```python
# Check both User and Teacher tables
user = User.query.filter_by(email=email).first()
teacher = Teacher.query.filter_by(email=email).first()

# Determine if the email belongs to a Teacher or a User
account = teacher if teacher else user
``` 

However, if we take a look at lines 64 to 84:
```python
role = request.form.get("role")
...
if role == "teacher":
    user_email = Teacher.query.filter_by(email=email).first()
    user_name = Teacher.query.filter_by(username=username).first()
else:
    user_email = User.query.filter_by(email=email).first()
    user_name = User.query.filter_by(username=username).first()
...
if user_name:
    flash("This username already exists.", "register_error")
    return redirect(url_for("auth.register"))
...
if user_email:
    flash("This email already exists.", "register_error")
    return redirect(url_for("auth.register"))   
``` 
This code shows that when a consumer signs up under some arbitrary username A and email B, the auth system checks for duplicate usernames/emails among only other teachers when signing up as a teacher, or only other students when signing up as a student. Thus, an actor/two separate actors could accidentally sign up with the same exact email A and password B under two accounts—one student, and one teacher. Problem is, given the way the if statement at line 31 is structured, the auth system will always choose only one account role over the other, and it always has a bias towards choosing teachers. As such, when logging in with email A and password B, the system will never log into the student's account and always access the teacher's. 

If you're going for a "can be both" approach and want the same email to have a teacher and student role, then we should probably add a similar radio role select button/use separate pages for login. Otherwise, you should probably check for duplicate emails/usernames within both the User and Teacher tables in lines 64  to 84, no matter the role.

alexchen2 commented:
For the second topic, this initially came to mind due to bugs with logging in as a Teacher; Flask-Login requires any entity inheriting `UserMixin` to specifically have an `id` attribute as its primary key (not `tid`), and Teacher was also missing an extra `failed_signin_attempts` field from User that was needed in `auth.py`. These fixes are pretty simple to fix by altering the Teacher table, but the User and Teacher tables have so many similarities between each other at this point that it feels inefficient and a waste of space to have them as separate tables. Wouldn't it just be better to have a main User table with shared attributes (name, hashed_password, email, etc.), then create ISA specializations for each role that inherit from the parent User table (Student having grade, points, etc; Teacher having yrs_experience, specialization, etc.)? This way, not only is our code more concise, but we wouldn't have to constantly create and manage duplicate columns between both tables taking up valuable storage. 

I do understand, however, that changing this DB structure this late into the project would most likely take a bit more time in fixing any resulting bugs with queries dependent on the old User model (with Student-specific attributes), so I understand if you aren't too keen on the idea.

angadk4 committed:
Implemented Requested changes in register.html/css/js, auth.py & models.py

angadk4 commented:
I have implemented all of the relevant changes recommended by Alex:
- Changes to _register.html, register.css, and register.js_ have been made exactly as recommended, with a few minor modifications. 
- Changes to _models.py_ have been implemented so that `teacher.tid` and all of its occurrrences are replaced by `teacher.id` to better fit the Flask model.
- In terms of _auth.py_, I have chosen to go with the approach of only allowing a user to either be a teacher or student. I modified registration logic to check for email and username in both Teacher and User tables to ensure no individuals have registered with these credentials in the past. I also modified the login logic to ensure there is no bias when checking for account type based on email queries.

Here are some other important things to note:
- I definitely agree that creating an ISA specialization to simplify these roles would simply the logic, however I have been told that RBAC is not yet a major feature of our application, and thus it's probably best to continue as is for now.
- **IMPORTANT:** If it is deemed that these new changes have been correctly implemented and this PR is ready for merge, please ensure to migrate the database after doing so. At the moment, changing the teacher `tid` to `id` is causing some issues with the login/registration logic. I am not well aware of the alembic system being used for migrations, so this will likely be better handled by someone else on the team.

alexchen2 commented on app/src/auth.py:
```diff
@@ -28,7 +28,14 @@ def login_post():
+        account = teacher
+    elif user and not teacher:
+        account = user
+    else:
```
Change the else to `elif user and teacher`; otherwise, if a user tries to log in with an invalid email, the proper login error prompt at line 51 will never be raised and this "multiple roles" one will be shown instead.

alexchen2 reviewed (changes requested):
Looks good to me for the most part—there are just some small minor changes left that need to be made.

In addition to the auth.py comment below, add this entry within the `errorMapping` dictionary in login.js at line 10:
```python
const errorMapping = {
    ...
    "The email address is associated with multiple roles. Please contact support." : "err-email"
    ...
}
```
This makes it so that the page actually detects the flashed message error being made and renders it properly under the appropriate textbox.

angadk4 committed:
error mapping for teacher/user

angadk4 commented:
Added changes to properly account for an error when email associated with both teacher & user.

alexchen2 commented on app/src/auth.py:
```diff
@@ -32,7 +32,7 @@ def login_post():
         account = teacher
     elif user and not teacher:
         account = user
-    else:
+    elif user and teacher:
         flash('The email address is associated with multiple roles. Please contact support.', 'login_error')
         return redirect(url_for('auth.login'))
```
Before this if statement (line 29-ish), add a line instantiating account such that `account = None`. Otherwise, if none of the conditionals here are triggered, the server will crash at the line 41 conditional due to the variable not existing before being used.

angadk4 replied:
Sounds good, I've made the change.

alexchen2 reviewed (changes requested):
Small bug that needs to be fixed from the last edit, but everything looks good to me otherwise. The PR should be ready to merge after this (although we do need to deal with the database migration issue somehow).

angadk4 committed:
account=None in auth.py

alexchen2 committed:
Deleted site.db (not needed, database in main repo is auto-generated)

The main repo filters out the site.db file in its .gitignore, and it doesn't contain one in an "instance/" path; this is probably since the file is auto-generated locally upon starting the server, so there's no need to have push this one to the repo.

alexchen2 committed:
Merge branch 'role-based' of https://github.com/angadk4/Gamified-Learning-Platform into pr/117
